File: Programmers Level1/키패드누르기.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

string solution(vector<int> numbers, string hand) {
	string answer = "";
	int rpos=12, lpos = 10;
	int llength, rlength;
	for (int i = 0; i < numbers.size(); i++) {
		if (numbers[i] == 1 || numbers[i] == 4 || numbers[i] == 7) {
			answer += "L";
			lpos = numbers[i];
		}
		else if (numbers[i] == 3 || numbers[i] == 6 || numbers[i] == 9) {
			answer += "R";
			rpos = numbers[i];

		}
		else if(numbers[i] == 2 || numbers[i] == 5 || numbers[i] == 8 || numbers[i] == 0) {
			if (lpos == 1 && numbers[i] == 2) {
				llength = 1;
			}
			else if (lpos == 1 && numbers[i] == 5) {
				llength = 2;
			}
			else if (lpos == 1 && numbers[i] == 8) {
				llength = 3;
			}
			else if (lpos == 1 && numbers[i] == 0) {
				llength = 4;
			}
			else if (lpos == 4 && numbers[i] == 2) {
				llength = 2;
			}
			else if (lpos == 4 && numbers[i] == 5) {
				llength = 1;
			}
			else if (lpos == 4 && numbers[i] == 8) {
				llength = 2;
			}
			else if (lpos == 4 && numbers[i] == 0) {
				llength = 3;
			}
			else if (lpos == 7 && numbers[i] == 2) {
				llength = 3;
			}
			else if (lpos == 7 && numbers[i] == 5) {
				llength = 2;
			}
			else if (lpos == 7 && numbers[i] == 8) {
				llength = 1;
			}
			else if (lpos == 7 && numbers[i] == 0) {
				llength = 2;
			}
			else if (lpos == 10 && numbers[i] == 2) {
				llength = 4;
			}
			else if (lpos == 10 && numbers[i] == 5) {
				llength = 3;
			}
			else if (lpos == 10 && numbers[i] == 8) {
				llength = 2;
			}
			else if (lpos == 10 && numbers[i] == 0) {
				llength = 1;
			}
			// 오른손
			if (rpos == 3 && numbers[i] == 2) {
				rlength = 1;
			}
			else if (rpos == 3 && numbers[i] == 5) {
				rlength = 2;
			}
			else if (rpos == 3 && numbers[i] == 8) {
				rlength = 3;
			}
			else if (rpos == 3 && numbers[i] == 0) {
				rlength = 4;
			}
			else if (rpos == 6 && numbers[i] == 2) {
				rlength = 2;
			}
			else if (rpos == 6 && numbers[i] == 5) {
				rlength = 1;
			}
			else if (rpos == 6 && numbers[i] == 8) {
				rlength = 2;
			}
			else if (rpos == 6 && numbers[i] == 0) {
				rlength = 3;
			}
			else if (rpos == 9 && numbers[i] == 2) {
				rlength = 3;
			}
			else if (rpos == 9 && numbers[i] == 5) {
				rlength = 2;
			}
			else if (rpos == 9 && numbers[i] == 8) {
				rlength = 1;
			}
			else if (rpos == 9 && numbers[i] == 0) {
				rlength = 2;
			}
			else if (rpos == 12 && numbers[i] == 2) {
				rlength = 4;
			}
			else if (rpos == 12 && numbers[i] == 5) {
				rlength = 3;
			}
			else if (rpos == 12 && numbers[i] == 8) {
				rlength = 2;
			}
			else if (rpos == 12 && numbers[i] == 0) {
				rlength = 1;
			}

			// 거리 계산
			if (llength < rlength) {
				answer += "L";
				lpos = numbers[i];
			}
			else if(llength > rlength) {
				answer += "R";
				rpos = numbers[i];
			}
			else if (llength == rlength) {
				
				if (hand == "right") {
					answer += "R";
					rpos = numbers[i];
				}

				else if (hand == "left") {
					answer += "L";
					lpos = numbers[i];
				}

			}
		}
			
	}
	return answer;
}
int main() {
	vector<int> a{ 7,0,8,2,8,3,1,5,7,6,2 };
	string h = "left";
	solution(a, h);
}

File: Programmers Level1 - Kotlin/키패드누르기.kt
import java.lang.StringBuilder
import kotlin.math.*

fun main() {
    solution(intArrayOf(0,0,1,1,5,7,9,9,3,0,1,2,3), "left")
}
fun solution(numbers: IntArray, hand: String): String {
    var answer = StringBuilder()
    var lnow: Int = 10
    var rnow: Int = 12

    var list = ArrayList<Int>()
    for(i in numbers.indices) {
        if(numbers[i] == 0) {
            list.add(11)
        }
        else {
            list.add(numbers[i])
        }
    }
    //                    1         2         3
    var xy = arrayOf(Pair(0,0), Pair(0,1), Pair(0,2),
                     Pair(1,0), Pair(1,1), Pair(1,2),
                     Pair(2,0), Pair(2,1), Pair(2,2),
                     Pair(3,0), Pair(3,1), Pair(3,2))

    list.forEach {

        if(it == 1 || it == 4 || it ==7) {
            answer.append("L")
            lnow = it
        }
        else if(it == 3 || it == 6 || it ==9) {
            answer.append("R")
            rnow = it
        }
        else {
         var length = distance(xy[it-1], xy[lnow - 1], xy[rnow - 1])
            if(length.first > length.second) {
                answer.append("R")
                rnow = it
            }
            else if(length.first < length.second) {
                answer.append("L")
                lnow = it
            }
            else {
                if(hand == "right") {
                    answer.append("R")
                    rnow = it
                }
                else {
                    answer.append("L")
                    lnow = it
                }
            }
        }

    }
    println(answer.toString())
    return answer.toString()
}

fun distance(destination: Pair<Int, Int>, l: Pair<Int, Int>, r: Pair<Int, Int>): Pair<Int, Int> {
    var lLength = abs(destination.first - l.first) + abs(destination.second - l.second)
    var rLength = abs(destination.first - r.first) + abs(destination.second - r.second)
    var length = Pair(lLength, rLength)
    return length
}
